Realm/ObjectStore/collection_notifications.hpp

```diff
@@ -0,0 +1,71 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_COLLECTION_NOTIFICATIONS_HPP
+#define REALM_COLLECTION_NOTIFICATIONS_HPP
+
+#include "index_set.hpp"
+#include "util/atomic_shared_ptr.hpp"
+
+#include <exception>
+#include <functional>
+#include <memory>
+#include <vector>
+
+namespace realm {
+namespace _impl {
+    class BackgroundCollection;
+}
+
+// A token which keeps an asynchronous query alive
+struct NotificationToken {
+    NotificationToken() = default;
+    NotificationToken(std::shared_ptr<_impl::BackgroundCollection> query, size_t token);
+    ~NotificationToken();
+
+    NotificationToken(NotificationToken&&);
+    NotificationToken& operator=(NotificationToken&&);
+
+    NotificationToken(NotificationToken const&) = delete;
+    NotificationToken& operator=(NotificationToken const&) = delete;
+
+private:
+    util::AtomicSharedPtr<_impl::BackgroundCollection> m_query;
```

Comment:
`m_query` doesn't seem like an appropriate name for this member given that the thing it holds is described as a collection.